### PR TITLE
feat: Querier: Split gRPC client into two.

### PR DESCRIPTION
**Pull request by DylanGuedes**
**What this PR does / why we need it**:
Split the gRPC client used by the querier into two, one for the communication with the scheduler, the other for communicating with the query-frontend.
- This change is retrocompatible: you don't have to change anything to keep existing behavior. 
- To configure the custom scheduler grpc client, you can use the new `query_scheduler_grpc_client` config or the new CLI flag `querier.scheduler-grpc-client`
- If you'd like to configure your frontend grpc client using a better named section, you can use the new `query_frontend_grpc_client` instead of the old `grpc_client_config`. Just make sure you don't use both at the same time, it will result in an error.

This work is necessary for configuring custom behavior between `querier<->scheduler` vs `querier<->frontend`. A use case is configuring mTLS when a different certificate is used by queriers, schedulers and frontends. You can only configure a single `server_name` with our current setup, making it impossible.

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Special notes for your reviewer**:

**Checklist**
- [ ] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [ ] Tests updated
- [ ] Title matches the required conventional commits format, see [here](https://www.conventionalcommits.org/en/v1.0.0/)
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

### CODE REVIEW

File: docs/sources/shared/configuration.md [hunk 1]
```diff
@@ -3557,10 +3557,23 @@ The `frontend_worker` configures the worker - running within the Loki querier -
 # CLI flag: -querier.id
 [id: <string> | default = ""]
 
-# The grpc_client block configures the gRPC client used to communicate between a
-# client and server component in Loki.
+# Configures the gRPC client used to communicate between the querier and the
```
Comment (salvacorts) [inline docs/sources/shared/configuration.md:3560]: I'd reword this the other way around: first what;s the common use-case, later the note about use_separated_clients being true.

File: pkg/querier/worker/worker.go [hunk 2]
```diff
@@ -92,9 +101,19 @@ type querierWorker struct {
...
 
 	metrics *Metrics
+
+	grpcClientConfig      grpcclient.Config
+	maxConcurrentRequests int
 }
 
 func NewQuerierWorker(cfg Config, rng ring.ReadRing, handler RequestHandler, logger log.Logger, reg prometheus.Registerer, codec RequestCodec) (services.Service, error) {
+	if !cfg.UseSeparatedGRPCClients {
+		level.Warn(logger).Log("msg", "Using the same client for frontend and scheduler. This is deprecated and will be removed in the future. Please use separated clients for frontend and scheduler.")
```
Comment (salvacorts) [inline pkg/querier/worker/worker.go:111]: IF this is deprecated, we should add a note to the docs.

Comment (salvacorts) [reply pkg/querier/worker/worker.go:111]: Are we sure we want to deprecate this? IIUC 99.% of the users will only need one grpc client. So we will require them to configure two identical grpc configs one for the frontend and another for the scheduler?

Comment (DylanGuedes) [reply pkg/querier/worker/worker.go]: great point. https://github.com/grafana/loki/pull/12726/commits/35820954f379ea91a62b369a88615125b9d20ed9 reverts it.

File: docs/sources/shared/configuration.md [hunk 3]
```diff
@@ -5329,83 +5328,83 @@ The `swift_storage_config` block configures the connection to OpenStack Object S
 
 ```yaml
 # OpenStack Swift authentication API version. 0 to autodetect.
-# CLI flag: -<prefix>.swift.auth-version
+# CLI flag: -swift.auth-version
```
Comment (salvacorts) [inline docs/sources/shared/configuration.md:5331]: I think these shouldn't be here.

**Review by salvacorts — APPROVED** [review]
LGTM! Thank you Dylan :)

File: docs/sources/shared/configuration.md [hunk 4]
```diff
@@ -543,19 +543,19 @@ The `alibabacloud_storage_config` block configures the connection to Alibaba Clo
 
 ```yaml
 # Name of OSS bucket.
-# CLI flag: -common.storage.oss.bucketname
```
Comment (DylanGuedes) [inline docs/sources/shared/configuration.md:546]: this flag docs was bugged this whole time
